### PR TITLE
Linux support

### LINKED ISSUE #10: Linux port
Hi. I'm asking for Linux port for this project. It runs terribly under Wine, and Unity gives me compilation errors when I try to build it (I'm n00b in C#, so I don't know what's going on).

### PR DESCRIPTION
**Pull request by bmwalters**
Fixes #10 

Each change is split into a separate commit to aid review.

## Changes required for Linux support
- Implement Linux native message box, file explorer, and window handle
  - Uses Xlib to get the X11 window handle
  - Uses GTK to display message box and file explorer
  - Fixed Linux file explorer bugs such as opening behind the game window
- Implement OGG encoding on Linux using FFmpeg
  - Uses system's FFmpeg
  - BASS' EncoderOGG is not available on Linux and mac due to its use of oggenc2.exe
  - FFmpeg is not required in Windows builds
- Use system SDL libraries
- Add BuildLinux option to editor build commands
- Update README.md with build instructions and cleanup

Lastly, I added an Arch Linux `PKGBUILD` file that can be used by the Arch Linux `makepkg` and `pacman` package manager to automatically (mostly...) build and install Moonscraper from the command line.

## Other changes included in this PR
- Log an error message when audio fails to load due to a BASS error
  - The application would hang when I tried to open an ogg file with embedded album art. Now it displays a nice error message.
- Overwrite existing files when exporting a song (after asking the user)
- Implement reading midi charts with non-ogg audio
- Allow BASS audio to work in the Unity editor
  - As part of this I removed the `BASS_AUDIO` Scripting Define Symbol. This causes the diff in `PlayerSettings.asset`. This would be easier to read if #33 was merged, but it doesn't really matter.

### CODE REVIEW

**Comment by bmwalters** [comment]
Annoyingly [GitHub displays the commits out of order](https://docs.github.com/en/github/committing-changes-to-your-project/why-are-my-commits-in-the-wrong-order) in the pull request view. From the command line the order is correct.

cc @dominuick @jedenastka @FireFox2000000

File: README.md [hunk 1]
```diff
@@ -1,30 +1,32 @@
-### THIS IS NOT THE APPLICATION PROGRAM, THESE ARE THE SOURCE FILES. ###
-### If you are looking download Moonscraper Chart Editor please download from ###
-https://github.com/FireFox2000000/Moonscraper-Chart-Editor/releases
-### Or visit the GOOGLE DRIVE link in the description of the Youtube trailer video. ###
+Moonscraper Chart Editor, a song editor for Guitar Hero style rhythm games made in Unity 2018.3.0f2
```
Comment (FireFox2000000) [inline README.md:1]: Version number is wrong, we're up to Unity 2018.4.23f1

File: README.md [hunk 2]
```diff
@@ -1,30 +1,32 @@
...
+> If you are looking to download Moonscraper Chart Editor please see the
+> [releases page](https://github.com/FireFox2000000/Moonscraper-Chart-Editor/releases).
 
-### WARNING ###
+## Compiling from Source
+Follow the instructions below for your desired platform to build and run from source.
 
-This project contains licensed assets such as Un4seen's "bass", "Bass.Net" and several icons. Please view the appropriate attribution documentation and licensing information if you intend to use these resources. 
+### All Platforms
+1. Download and install Unity 2018.3.0f2
```
Comment (FireFox2000000) [inline README.md:12]: Version number is wrong, we're up to Unity 2018.4.23f1

**Comment by bmwalters** [comment]
> Removal of #if BASS_AUDIO in the audio manager code seems unnecessary? BASS_AUDIO should be defined in unity scripting defines to allow audio to play in the editor. Optional BASS_AUDIO was define in case the codebase ever needed to swap audio libraries.
> 
> (Commit [16f8478](https://github.com/FireFox2000000/Moonscraper-Chart-Editor/commit/16f84789cb374638fa95703deccc9133b608cf22))

**Comment by bmwalters** [comment]
@FireFox2000000 Sorry about the problems in that commit. I believe I've addressed the issues you found if you would like to take another look.

**Comment by bmwalters** [comment]
@FireFox2000000 Looks like X11.dll needs the "AnyOS" platform selected in its Unity Editor platform settings. Not sure why, but I made this change and I was able to build from Windows then run the result on Linux.

**Comment by FireFox2000000** [comment]
Successfully built. Currently finding and awaiting on volunteers to verify the build and to run through specific tests. If you would like to run through the tests yourself you can find the link at https://drive.google.com/file/d/1fWGoR12bxdnYGXEhZunQsmaKioFLfE1G/view?usp=sharing

When submitting test results please specify the exact version of Linux the test ran on and any output_log.txt as specified by the tests.

**Comment by bmwalters** [comment]
I ran through the tests @FireFox2000000 

Notes:
- Instead of `output_log.txt` I had a file named `Player.log`.
- After downloading and extracting the rar I had to run `chmod +x "Moonscraper Chart Editor.x86_64"` to make the program executable. I had to do this when I built from Windows, but not when I built from Linux.
- All tests were ran using Moonscraper from your rar file with the exception of Native GUI 2 and Custom Resources 1, since those tests uncovered the following issues which I fixed:
  - Choosing "Yes" when asked if the user wants to save before exiting resulted in an editor hang.
  - I needed to use X11 instead of Wayland to record the video for Custom Resources. This uncovered an issue where the window handle failed to be obtained.
- All tests passed after those two fixes.

System:
- Arch Linux x86_64, kernel 5.7.7-arch1-1, GNOME Wayland (running Moonscraper under XWayland)

<details>

<summary>Expand to see detailed results</summary>

| Feature | Test | Result |
| ---- | ----- | ------- |
| Native GUI | -Boot up Moonscraper Chart Editor<br>-Upon loading into the main game scene and with Moonscraper focused, the application window title should read<br>"Untitled - Guitar - Expert - Moonscraper Chart Editor v1.3.5"  | Success<br>screenshot |
| Native GUI | -Follow on from test 1.<br>-Navigate to File->Load<br>-Open file dialog should be present.<br>-Attempting to click back onto the application with the dialog open should be blocked and focus should return<br>to the open file dialog. <br><br>-Clicking "Cancel" should return back to the application with no noticable changes<br><br>-Navigate to Test Files/ogg test/Victory Solo X - HOPO [Y].chart, click the file and select "Open".<br>File should successfully load and chart populated. | Success<br>*Note:* Clicking on the application window while the file browser is open displays an "Application is not responding" dialog. This can be ignored and is not uncommon. screenshot of file browser<br>screenshot of opened chart |
| Native GUI| -Follow on from test 1.<br>-Select the note tool and add a note to the chart. <br>-The application window title should read<br>"Untitled * - Guitar - Expert - Moonscraper Chart Editor v1.3.5"  (Note the arterisk)<br><br>-Attempt to close the application. On windows this would normally be done by hitting the "X" in the <br>top-right corner of the application<br>-Popup dialog with the options "Yes", "No" and "Cancel" should appear that reads<br>"Warning"<br>"Do you want to save unsaved changes?"<br><br>-Attempting to click back onto the application with the dialog open should be blocked and focus should return<br>to the popup dialog. <br><br>-Click the "Cancel" button. The application should return back to the application with no noticable changes<br><br>-Attempt to close again and select either "Yes". The application should open up to a save file dialog.<br><br>-Select "Cancel". The application should return back to the application with no noticable changes<br><br>-Attempt to close again and select "No". The application should close successfully. | Success<br>*Note:* Instead of opening a save dialog when "Yes" is pressed, it just saves over the opened chart file.<br>screenshot of added note<br>screenshot of close confirmation |
| Audio | -Boot up Moonscraper Chart Editor<br>-Open up the chart file located under Test Files\ogg test\Victory Solo X - HOPO [Y].chart<br>-Press spacebar to play the chart. Audio should be present. | Success<br>*Note:* I had to use "Song Properties" to select "VICTORY SOLO X.ogg". |
| Audio | -Boot up Moonscraper Chart Editor<br>-Open up the chart file located under Test Files\mp3 test\Destruction Armeggadon (Underchart) [Y].chart<br>-Press spacebar to play the chart. Audio should be present. | Success |
| Audio | -Boot up Moonscraper Chart Editor<br>-Open up the chart file located under Test Files\wav test\notes.chart<br>-Press spacebar to play the chart. Audio should be present once the first note gets hit | Success |
| Audio | -Boot up Moonscraper Chart Editor<br>-Open up the chart file located under Test Files\mp3 test\Destruction Armeggadon (Underchart) [Y].chart<br>-Navigate to File->Export<br>-Enable the "Clone Hero Package" option.<br>-"Select Folder" file dialog option should be present<br>-Attempt to export under the "Test Files" folder.<br>-Under Test Files, a folder named "Destruction Armaggedon (Underchart)" should have been created with<br>the following files present:<br>-notes.chart<br>-song.ini<br>-song.ogg | Success<br>screenshot of ogg encoding<br>screenshot of exported package |
| Networking | -Boot up Moonscraper Chart Editor<br>-Navigate to Help->Check For Updates. Wait for the menu to load<br>-Menu should read "Moonscraper Chart Editor is up to date. The current version is v1.3.5" | Success<br>screenshot of update check<br>[Networking Test 1.txt](https://github.com/FireFox2000000/Moonscraper-Chart-Editor/files/4875789/Networking.Test.1.txt) |
| Networking | -Boot up Moonscraper Chart Editor<br>-Navigate to Help->Contact.<br>-Click on the "Github" button.<br>-Default internet browser should open and load up the webpage<br>"https://github.com/FireFox2000000/Moonscraper-Chart-Editor" | Success |
| Custom Resources | -Copy the Folder "Test Files\Custom Resources" into "Moonscraper Chart Editor v1.3.5 Linux (Universal)"<br>Overwrite any files as prompted.<br>-Boot up Moonscraper Chart Editor<br>-Note that notes and other assets should have successfully loaded changed to those in the custom resources folder<br>-Either load up a chart or place some notes.<br>-Enable the Clap option<br>-Play song and note that clap sound should have change to that found in the custom resources folder. | Success<br>*Note:* I ran this test under X11 so I could record my screen.<br>[custom_resources.mp4.zip](https://github.com/FireFox2000000/Moonscraper-Chart-Editor/files/4875903/custom_resources.mp4.zip) |

</details>

**Comment by FireFox2000000** [comment]
For Native GUI Test 3, you noted 
> Instead of opening a save dialog when "Yes" is pressed, it just saves over the opened chart file.
From the screenshot it appears that a chart was loaded in when the test was ran. It is expected that the save dialog should appear upon making an edit to a blank chart, or a chart that was not loaded from a .chart file.

**Comment by FireFox2000000** [comment]
For Audio Test 1, you noted

> I had to use "Song Properties" to select "VICTORY SOLO X.ogg".

Which indicates that when you loaded the chart file, the audio failed to load at the same time despite the reference to the audio file being present in the file
`MusicStream = "VICTORY SOLO X.ogg"`
This would be a problem if it's not loading corresponding audio files when you load up charts. Would you be able to reconfirm if that is an issue?

**Comment by Mario64iscool2** [comment]
I'm having issues locating Output.txt OR player.log

**Comment by Mario64iscool2** [comment]
Using Moonscraper on Linux feels mostly like using it on Windows, except for one thing: The scroll is inverted and much less sensitive.

**Comment by Mario64iscool2** [comment]
### Results of Testing:

Native GUI:
   Application Window title should read "Untitled - Guitar - Expert - Moonscraper Chart Editor v1.3.5" Success

  Attempting to click back onto the application with the dialog open should be blocked and focus should return to the popup
Focus does not return to the file popup. Input, however, is blocked.

Clicking "Cancel" Should return back to the application with no noticeable changes.
Success

Navigate to Test Files/ogg test/Victory Solo X -HOPO [Y}. chart, file should load with notes populated.
Success'

Add a note, the window title should now show that you have unsaved changes
Success

Attempt to close the app and the "Do you want to save unsaved changes" dialog should appear
Success

Attempting to click back onto the application with the dialog open should be blocked and focus should return to the popup.
Attempting to click back onto the application with the dialog open should be blocked and focus should return to the popup
Focus does not return to the file popup. Input, however, is blocked.

Click cancel. The app should return to normal functionality.
Success

Attempt to close again and select "Yes". the app should open to a save dialog
Success

Select cancel and the app returns to normal
Success

Attempt to close again and select no. The app should successfully close.
Success

### Audio

-Boot up Moonscraper Chart Editor
      -Open up the chart file located under Test Files\ogg test\Victory Solo X - HOPO [Y].chart
      -Press spacebar to play the chart. Audio should be present. 
          Success

-Boot up Moonscraper Chart Editor
      -Open up the chart file located under Test Files\mp3 test\Destruction Armeggadon (Underchart) [Y].chart
      -Press spacebar to play the chart. Audio should be present. 
          (Success)

-Boot up Moonscraper Chart Editor
      -Open up the chart file located under Test Files\wav test\notes.chart
      -Press spacebar to play the chart. Audio should be present once the first note gets hit. 
          (Success)

-Boot up Moonscraper Chart Editor
      -Open up the chart file located under Test Files\mp3 test\Destruction Armeggadon (Underchart) [Y].chart
      -Navigate to File->Export
      -Enable the "Clone Hero Package" option.
      -"Select Folder" file dialog option should be present
      -Attempt to export under the "Test Files" folder.
      -Under Test Files, a folder named "Destruction Armaggedon (Underchart)" should have been created with
      the following files present:
        -notes.chart
        -song.ini
        -song.ogg
Success

### Networking
-Boot up Moonscraper Chart Editor
      -Navigate to Help->Check For Updates. Wait for the menu to load
      -Menu should read "Moonscraper Chart Editor is up to date. The current version is v1.3.5"
          (Success)

      -Boot up Moonscraper Chart Editor
      -Navigate to Help->Contact.
      -Click on the "Github" button.
      -Default internet browser should open and load up the webpage
      "https://github.com/FireFox2000000/Moonscraper-Chart-Editor"
Success
          
### Custom Resources

-Copy the Folder "Test Files\Custom Resources" into "Moonscraper Chart Editor v1.3.5 Linux (Universal)"
    Overwrite any files as prompted.
      -Boot up Moonscraper Chart Editor
      -Note that notes and other assets should have successfully loaded changed to those in the custom resources folder
        (Success)

-Either load up a chart or place some notes.
      -Enable the Clap option
      -Play song and note that clap sound should have change to that found in the custom resources folder.
(Success)
[Networking Test 1.txt](https://github.com/FireFox2000000/Moonscraper-Chart-Editor/files/4880689/Networking.Test.1.txt)

Manjaro Linux KDE x86_64
Kernel 5.6.16-1-MANJARO

**Comment by bmwalters** [comment]
Re ogg file audio:
My .chart file didn't have that line. I don't know what happened originally but I just redownloaded the test files and it worked great this time.

Re save dialog:
You're right. & when I test with a blank chart it does work as expected.

> For commit [5264524](https://github.com/FireFox2000000/Moonscraper-Chart-Editor/commit/52645245a7ab6dde585ac9339d45a902433396f7)
> @bmwalters
> 
> If what you say is correct, wouldn't moving FinaliseQuit() which has the blocking in it, from the
> `Application.quitting `
> actions over to within
> 
> ```
> if (quittingCheckPassed)
> {
>     Application.Quit();
>    // Something like FinaliseQuit();
> }
> ```
> 
> be the better solution? That way the waiting for save blocking will run on the main thread while the actual save task can run on the UI thread if it wants to without issues.

I think that would work in this case since we could get rid of `while (isSaving) ;` within `FinaliseQuit` because we could check `if (quittingCheckPassed && !isSaving)` every frame using `Update`. Though it wouldn't solve the problem for cases like `New()` where `EditCheck` is called then it immediately blocks on `while (isSaving) ;` which could deadlock in theory.

**Comment by FireFox2000000** [comment]
One other thing I forgot to include in the tests were inputs, since that's based on the SDL libs. Might be worth double checking with some random controllers that the rebinding menu and controls mapped to those are still all good too.

**Comment by bmwalters** [comment]
I've also experienced the scroll behavior described.

I tested using my custom Arduino-converted WT guitar (implemented as a generic USB HID gamepad) and binding controls worked, along with fretting and strumming.

**Comment by FireFox2000000** [comment]
Mouse scroll stuff was a bit hacky anyway. I'll see if I can replace the scroll input using SDL instead of Unity to see if that can help with the input consistency. 

@bmwalters How do you want to be credited for the Linux build? Just "Bradley Walters" with a link to your Github page or something like that?

**Comment by bmwalters** [comment]
Just [bmwalters](https://github.com/bmwalters) with a GitHub link is fine, thanks.

**Comment by FireFox2000000** [comment]
@bmwalters 
Hopefully this build should have the scroll issue fixed- https://drive.google.com/file/d/1LLNYf0-XgbQs4zEfJ3v0PXUoOf1TYryr/view?usp=sharing

Also worth noting that building Windows breaks unless I toggle off the "Auto Reference" flag on Bass.Net.Linux, and turn it back on for the Linux builds. I don't know why, I suspect it's a bug with Unity not filtering by platform correctly.

**Comment by bmwalters** [comment]
@FireFox2000000 Funnily enough I just encountered the same problem when working on macOS support. You can fix it by toggling every single checkbox in the meta file for Bass.Net.Linux then committing the result. Apparently there's a difference in Unity between "checkbox has never been pressed" and "checkbox has been toggled on then off".

I will let you know how the scrolling is on that build.

**Comment by FireFox2000000** [comment]
@bmwalters Yeah that appears to have worked, ty. Committed and pushed.

**Pull request merged**